### PR TITLE
Partial data import

### PR DESCRIPTION
**Pull request by stmichael**
I started by decoupling the lookup tables from the definitions. It's not finished yet. There is still code that uses lookups on definitions.

@senny what do you think about it so far?

### CODE REVIEW

**Comment by stmichael** [comment]
Also references to lookup tables are still hold on the definitions. If we really want to make the lookup an independent feature, we have to put it somewhere else.

File: lib/data-import/definition/script.rb [hunk 1]
```diff
@@ -3,7 +3,22 @@ class Definition
...
+        super(name, source_database, target_database)
+        @lookup = lookup
+      end
+
+      def lookup_for(*args)
+        @lookup.lookup_for(*args)
+      end
+
+      def row_imported(*args)
+        @lookup.row_imported(*args)
```
Comment (senny) [inline lib/data-import/definition/script.rb]: We split out the Lookup to have a separate entity, that we can reason about. If you think in the terms of a LookupTable it does not care when a row was imported. The method on the lookup should have a more intention revealing name.

File: lib/data-import/definition/lookup.rb [hunk 2]
```diff
@@ -1,9 +1,11 @@
 module DataImport
   class Definition
-    module Lookup
-      def initialize(*args)
+    class Lookup
```
Comment (senny) [inline lib/data-import/definition/lookup.rb]: A Lookup is more an action than a thing. Since this class holds a table of lookup data I think we should rename it to something more expressive:
- LookupTable
- IdDictionary
- MappingTable
- you get the idea ;)

File: lib/data-import/definition/script.rb [hunk 3]
```diff
@@ -3,7 +3,22 @@ class Definition
     class Script < Definition
       attr_accessor :body
 
-      include Lookup
+      def initialize(name, source_database, target_database, lookup)
```
Comment (senny) [inline lib/data-import/definition/script.rb]: I think the `Lookup` should not be directly coupled to one definition. I think we should have one LookupTable per mapping. If a definition has 2 LookupTables we should have two instances somewhere. @stmichael what do you think?

Comment (stmichael) [inline lib/data-import/definition/script.rb]: What do you mean by "I think we should have one LookupTable per mapping"? In terms of data-import, do you actually mean mapping or definition? Or maybe something else?

Let's call the lookups id dictionaries for the moment to be more precise.

I thought of saving the id dictionaries in the execution plan in a similar way as the definitions. If a definition is created the dsl also creates an id dictionary with the same name as the definition.

`lookup` passed to the constructor of the script and import definition is just to ensure backwards compatibility. As soon as we release a stable version, this can be dropped.

Comment (senny) [inline lib/data-import/definition/script.rb]: The current `Lookup` class has a variable called `@lookup_tables = {}` I think every table (or dictionary) should be a sperate instance. You can then ask it to `#store` a new entry and to `#lookup` an ID by value. The glue logic needed, to assign lookup tables to attributes and definitions should not be part of that `Dictionary` instance. You can den define a simple "DictionarySerializer` interface to store those instances on disk, in memory or wherever you want.

**Comment by stmichael** [comment]
I created an id mapping container which holds all id dictionaries. The container is held by the execution plan. I'm not happy the way it is used because I have to pass it to several initializations. @senny any idea on how to change that? May the container is in the wrong place.

The syntax to use the id mappings in the blocks is now:

``` ruby
id_mapping_for('definition name', 'id lookup name').lookup(old_value)
id_mapping_for('definition name', 'id lookup name').add(old_value, new_value)
```

File: lib/data-import/definition/dictionary.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
+module DataImport
+  class Definition
+    class Dictionary
```
Comment (senny) [inline lib/data-import/definition/dictionary.rb]: I think we can move this class outside the `Definition` namespace.

**Comment by senny** [comment]
@stmichael I think the IdMappingContainer belongs to the ExecutionPlan. I think the DSL objects should always have access to the plan, which is being built. They are coupled to the plan very tightly anyways because their only intention is to build the plan. So I suggest you pass along the whole plan in place of only the mapping container. I also think you should bridge the methods to add a new dictionary on the ExecutionPlan so that you don't need to reach through the object to grab the container.

**Comment by stmichael** [comment]
I updated the branch with the discussed changes. @senny If you agree I will go on to the implementation of the partial import

**Comment by stmichael** [comment]
This is a first version of a partial import. What do you think about it?

**Comment by stmichael** [comment]
The next step would be to extract the loading and storage of the settings into a separate class to keep it together. Also in the unit test of the partial import I directly write to the settings file as test setup. That's not flexible since it's closely tied to the implementation.

File: lib/data-import/definition.rb [hunk 5]
```diff
@@ -10,13 +10,53 @@ class Definition
...
+              CaseIgnoringDictionary.new
+            else
+              Dictionary.new
+            end
+        @id_mapping_container.add_dictionary(name, mapping_name, attribute, d)
+      end
+    end
+
+    def row_imported(id, row)
+      warn "[DEPRECATION] `definition(...).row_importer(...)` is deprecated and will be removed in later versions! Use `id_mapping_for(...).add(...)` instead.\n#{caller[0]}"
```
Comment (senny) [inline lib/data-import/definition.rb]: `row_importer` typo

File: lib/data-import/migration_strategy.rb [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+      @progress_reporter = progress_reporter
+    end
+
+    def run(options = {})
+      dependency_resolver = DependencyResolver.new(@plan)
+      @resolved_plan = dependency_resolver.resolve(:run_only => options[:only])
+      before_run(options)
+
+      @resolved_plan.definitions.each do |definition|
+        next if skip_definition?(definition)
```
Comment (senny) [inline lib/data-import/migration_strategy.rb]: why not itereate only over the definitions you want to import? This way the `MigrationStrategy` subclasses gain more flexibility.

File: lib/data-import/full_migration.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+module DataImport
+  class FullMigration < MigrationStrategy
+    def before_run(options)
+      @completed_definitions = []
```
Comment (senny) [inline lib/data-import/full_migration.rb]: instance variable dependencies over a subclass hierarchy should be prevented whenever possible. This makes the code brittle and hard to follow.

**Comment by stmichael** [comment]
What do you think about it now?

File: spec/unit/data-import/partial_migration_spec.rb [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    let(:mock_settings_store) do
+      Class.new do
+        def load
+          {}
+        end
+
+        def save(data)
+        end
+      end
+    end
```
Comment (senny) [inline spec/unit/data-import/partial_migration_spec.rb:28]: Why don't we use rspec-mocks? Is it not relevant if save was called?

File: spec/unit/data-import/partial_migration_spec.rb [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    let(:mock_settings_store) do
+      Class.new do
+        def load
+          {:completed_definitions => ['Animals']}
+        end
+
+        def save(data)
+        end
+      end
+    end
```
Comment (senny) [inline spec/unit/data-import/partial_migration_spec.rb:72]: Same as above

**Comment by senny** [comment]
@stmichael I'm a little short on review time at the moment. I have some general spots I'm not that happy with but don't have prepared suggestions how to improve:
1. The Naming seems a bit off. (eg. SettingsStore does not store settings but more progress related data, also the IdMappingsXXXX part does not communicate clearly to me. => very subjective comment)
2. I would not use Inheritance in the Migration Strategy but Composition.
3. I'm not sure if we want a two step process to lookup an ID, eg: `id_mapping_for(...).lookup(...)`. This was the source that we had to deprecate the complete API and I don't think we should do that in the future.
